### PR TITLE
Added basic multiline support based on stage-chains idea

### PR DESCRIPTION
**Pull request by pstibrany**
Added basic support for multiline stage.

It's configuration looks like this:

```yaml
  pipeline_stages:
    - multiline:
        firstline: '^START'
        max_wait_time: 1s
```

If line matches first line, then it is buffered and multiline stage waits for additional line until next 'firstline' or max wait time is elapsed.

This PR builds on PR #1375. PoC, don't merge.

### CODE REVIEW

File: pkg/logentry/stages/multiline.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	ErrEmptyMultilineStageConfig = "multiline stage config cannot be empty"
+	ErrMultilineNameRequired     = "multiline stage pipeline name can be omitted but cannot be an empty string"
+	ErrFirstLineRequired         = "firstLine regular expression required"
+)
+
+// MultilineConfig contains the configuration for a multilineStage
+type MultilineConfig struct {
+	PipelineName    *string `mapstructure:"pipeline_name"`
+	FirstLineRegexp string  `mapstructure:"firstline"`
+	MaxWait         string  `mapstructure:"max_wait_time"`
```
Comment (cyriltovena) [inline pkg/logentry/stages/multiline.go]: I think we should have a combination of max_wait_time and max_line. We can have this a second PR thought.

File: pkg/logentry/stages/pipeline.go [hunk 2]
```diff
@@ -108,14 +114,22 @@ func (p *Pipeline) Name() string {
 
 // Wrap implements EntryMiddleware
 func (p *Pipeline) Wrap(next api.EntryHandler) api.EntryHandler {
+	// start flushing every 100ms
+	go func() {
+		for range time.Tick(100 * time.Millisecond) {
```
Comment (cyriltovena) [inline pkg/logentry/stages/pipeline.go]: This needs a way to be stopped. You could use a context passed via the construct of the Pipeline.

File: pkg/logentry/stages/stage.go [hunk 3]
```diff
@@ -22,26 +22,44 @@ const (
...
+// to the next stage in the chain by calling chain.NextStage method.
+// If stage doesn't call next stage, no further processing is done.
 type Stage interface {
-	Process(labels model.LabelSet, extracted map[string]interface{}, time *time.Time, entry *string)
+	Process(labels model.LabelSet, extracted map[string]interface{}, time time.Time, entry string, chain StageChain)
 	Name() string
 }
 
+// FlushableStage is a stage that can be flushed.
+type FlushableStage interface {
```
Comment (cyriltovena) [inline pkg/logentry/stages/stage.go]: Can you elaborate on the comment, flush what ? when ? how ?

File: pkg/logentry/stages/pipeline.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/logentry/stages/pipeline.go]: or a close function that will cancel a context, I would still make this context from the original context so the cancellation above this context is propagated here.

**Comment by thomasdraebing** [comment]
I was able to use this pull request to successfully collect Java logs, including stacktraces as also discussed in #74. It would be awesome to have this feature in a release soon. Is there an ETA, when this will be merged?

Thanks,
Thomas

**Comment by pstibrany** [comment]
> I was able to use this pull request to successfully collect Java logs, including stacktraces as also discussed in #74. It would be awesome to have this feature in a release soon. Is there an ETA, when this will be merged?

Hi Thomas, it needs some more work before it's ready (like addressing Cyril's comments). Unfortunately I am currently busy with some other stuff, so I don't have time to finish it soon.

**Comment by cyriltovena** [comment]
Also we agreed, to get this feature in, so this is just a matter of time.

**Comment by thomasdraebing** [comment]
Thanks for the quick replies and thanks for working on this feature.
One thing I noticed is, that when looking at multiline logs in Grafana, the log will be shown as a single line in the logs panel or in the exploration view. However, the Loki Rest API seems to still contain the line breaks and tabs, so I guess this is due to Grafana omitting those.

**Comment by pstibrany** [comment]
> Thanks for the quick replies and thanks for working on this feature.
> One thing I noticed is, that when looking at multiline logs in Grafana, the log will be shown as a single line in the logs panel or in the exploration view. However, the Loki Rest API seems to still contain the line breaks and tabs, so I guess this is due to Grafana omitting those.

Yes. https://github.com/grafana/grafana/issues/20865

**Comment by RichiH** [comment]
Maybe not for this iteration of the feature, but there's network gear which outputs multiple log lines at the same timestamp with several distinct groups of logs. The only way to reliable deal with this is to be able to define and end string/glob/regex as well as a starting one.

**Comment by ahmed-lotfy** [comment]
Hi, any update about this feature ? we really need it!

**Comment by ricamz** [comment]
Hi.

Is it possible to match several lines with a pattern into a single line?

For example if I have the following lines:

<timestamp> 111 xxxxxxxxxxxxxxxxxxxxxxxx
<timestamp> 111 yyyyyyyyyyyyyyyyyyyyyyyy
<timestamp> 111 xxxxzzzzzzzzzzzzzzzzzzzzz
<timestamp> 111 yyyyyyyyyyyyyyyyyfsdfsdfsd sgfgdfgsd
<timestamp> 222 xxxxxxxxxxxxxxxxxxrwer werwerwerew
<timestamp> 222 yyyyyyyyyyywerwerwerwerwerwerwe

Join one line with the 111 lines and another line with the 222 lines.

Thanks
Ricardo

**Pull request closed without merging**